Allow TOV executables to use generic equations of state

isaaclegred opened the pull request:
## Proposed changes

This PR changes the structure of the EoS creation from input files and the structure of the TOV initial data for GRMHD to allow for use of generic equations of state implemented in spectre rather than just a polytropic equation of state.  This framework can be extended to allow any executable to use a generic equation of state in the input file.  

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

Currently the EoS is specified in the input file as a subset of the initial data, therefore changing the EoS to be generic only required changing the initial data.  In addition, most likely, the EoS used in the simulation should be made independent of the EoS used to generate the initial data and should be specified separately.  For example, it will become valuable to use 1-D EoSs in the initial data generation  while using a 2-D or 3-D EoS in the evolution.  This can be or can not be added to this pull request.  (Also need to fix merge conflicts with recent changes)

isaaclegred committed:
change TOV executables to use generic EquationOfState

isaaclegred commented:
@nilsdeppe and @ffoucart This pull request is functional, but there are some minutiae that need to be ironed out. This should allow a static-metric TOV evolution with arbitrary EoSs.  Essentially all that's required on the user's end is a change of input file.

nilsdeppe commented on src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp:
```diff
@@ -201,6 +201,29 @@ struct GhValenciaDivCleanDefaults {
                                      grmhd::ValenciaDivClean::Tags::TildeS<>,
                                      grmhd::ValenciaDivClean::Tags::TildeB<>>>>;
 
+  enum class Phase {
```
There's now a unified Phase class. Probably a bit of work to do when rebasing. Larry did this across a couple PRs

isaaclegred replied:
I haven't changed this yet maybe we can go over this at some point,  will it mainly entail removing code?

kidder replied:
just get rid of the `enum class Phase` and the `phase_name` function.

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/Tov.cpp:
```diff
@@ -99,6 +103,14 @@ void lindblom_rhs(
+    // This statement is triggered by one of the test examples,
+    // Test_Tov.cpp: void {anonymous}::test_rueter()
+    // Test_Tov.cpp: void {anonymous}::test_baumgarte_shapiro()
+    // not sure what to do about that
```
@kidder any idea what's happening here?

isaaclegred replied:
The naive answer seems to be that there is a place inside of the star where 2M/r > 1, this might be a distinction between isotropic radius and areal radius

nilsdeppe replied:
Ah yea, that's probably what it is. That the condition only applies in Schwarzschild-like coordinates

isaaclegred replied:
Hm I changed the assert statement to only run when the coordinates as Schwarzschild and I still get the error, it must be `test_baumgarte_shapiro` which is failing because it is the only of the pair that uses Schwarzschild coordinates.  I feel like based on the TOV solution there's no way somewhere inside the star should be able to have M/R > .5, but apparently here it does.  I also locally checked what the `radius_squared` value is at the point with M/R>.5 and it's apparently  0.745597 which is greater than the final value of .763^2, so I feel like something is definitely wrong.

nilsdeppe commented on src/PointwiseFunctions/Hydro/EquationsOfState/DarkEnergyFluid.hpp:
```diff
@@ -53,6 +53,13 @@ class DarkEnergyFluid : public EquationOfState<IsRelativistic, 2> {
+  std::unique_ptr<EquationOfState<IsRelativistic, 2>> get_clone()
+      const override {
+    auto clone = std::make_unique<DarkEnergyFluid<IsRelativistic>>(*this);
+    return std::unique_ptr<EquationOfState<IsRelativistic, 2>>(
```
If you wanted to you can inline the `std::make_unique...` into the `std::unique_ptr<...>(std::make_unique(...))`. Up to you, just keep the different implementations consistent :)

nilsdeppe commented on src/PointwiseFunctions/Hydro/EquationsOfState/DarkEnergyFluid.hpp:
```diff
@@ -53,6 +53,13 @@ class DarkEnergyFluid : public EquationOfState<IsRelativistic, 2> {
                 "Dark energy fluid equation of state only makes sense in a "
                 "relativistic setting.");
 
+  std::unique_ptr<EquationOfState<IsRelativistic, 2>> get_clone()
```
Could you add a test for the `get_clone` functions to the EOS tests? If there are some generic EOS tests then those could do that

nilsdeppe commented on src/PointwiseFunctions/Hydro/EquationsOfState/IdealFluid.hpp:
```diff
@@ -49,6 +49,12 @@ class IdealFluid : public EquationOfState<IsRelativistic, 2> {
   static constexpr bool is_relativistic = IsRelativistic;
 
+  std::unique_ptr<EquationOfState<IsRelativistic, 2>> get_clone()
+      const override {
```
move definition to cpp

nilsdeppe commented on src/PointwiseFunctions/Hydro/EquationsOfState/PolytropicFluid.hpp:
```diff
@@ -41,6 +41,13 @@ class PolytropicFluid : public EquationOfState<IsRelativistic, 1> {
 
+  std::unique_ptr<EquationOfState<IsRelativistic, 1>> get_clone()
+      const override {
+    auto clone = std::make_unique<PolytropicFluid<IsRelativistic>>(*this);
```
move to cpp

nilsdeppe commented on src/PointwiseFunctions/Hydro/EquationsOfState/Spectral.hpp:
```diff
@@ -57,6 +57,11 @@ class Spectral : public EquationOfState<true, 1> {
 
+  std::unique_ptr<EquationOfState<true, 1>> get_clone() const override {
+    auto clone = std::make_unique<Spectral>(*this);
+    return std::unique_ptr<EquationOfState<true, 1>>(std::move(clone));
```
move to cpp

nilsdeppe commented on src/PointwiseFunctions/Hydro/Tags.hpp:
```diff
@@ -211,4 +213,19 @@ struct MassFlux : db::SimpleTag {
   static std::string name() { return Frame::prefix<Fr>() + "MassFlux"; }
 };
 }  // namespace Tags
+namespace OptionTags {
```
Line before and then document the namespace similar to the `Tags` namespace

nilsdeppe commented on src/PointwiseFunctions/Hydro/Tags.hpp:
```diff
@@ -211,4 +213,19 @@ struct MassFlux : db::SimpleTag {
 };
 }  // namespace Tags
+namespace OptionTags {
+template <size_t ThermoDim = 0, bool IsRelativistic = true>
```
add doxygen comment

nilsdeppe commented on tests/Unit/Evolution/Initialization/Test_ConservativeSystem.cpp:
```diff
@@ -89,9 +91,10 @@ void check_primitives(std::true_type /*has_prims*/, const Runner& runner,
-            comp, typename Metavariables<Dim, true>::equation_of_state_tag>(
-            runner, 0) == 5);
+  // TODO :  no generic way to compare EoS
+  // CHECK(ActionTesting::get_databox_tag<
```
If you know the EOS you can do something like `dynamic_cast<const EquationOfState::PolytropicFluid&>(get_databox_tag...)`

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/Test_MagnetizedTovStar.cpp:
```diff
@@ -34,46 +36,68 @@ static_assert(
+      2,
+      0.04,
+      2500.0};
+  // const auto mag_tov = serialize_and_deserialize(mag_tov_original);
```
uncomment. We need to have things serializable

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/Test_MagnetizedTovStar.cpp:
```diff
@@ -34,46 +36,68 @@ static_assert(
+          2.28e-3,
+          std::make_unique<EquationsOfState::PolytropicFluid<true>>(100.0, 2.0),
+          TovCoordinates::Schwarzschild, 2, 0.04, 2500.0));
+  // TODO : Need a comparison operator for arbitrary equations of state.
```
This has been a bit of a contentious topic (can you/how should you compare abstract base classes). One way is to an `is_equal` virtual function that takes a `const EosBase&` argument, then have each derived EOS override the implementation, try to cast to itself, and if the cast works do the standard comparison. E.g.
```cpp
bool PolytropicFluid::is_equal(const Base& eos_base) const override {
  const auto& derived_ptr = dynamic_cast<const PolytropicFluid* const>(&eos_base);
  return derived_ptr != nullptr and *derived_ptr == *this;
}
```
I'm generally in favor of something like that since having some ability to do these types of comparisons is occasionally useful, but I'll bring it up at a spectre call (hopefully today)

isaaclegred replied:
Any news on this?  I feel like I tried this but something seemed difficult, maybe it had to do with the templated base class and the feeling that as it stands such a function would pragmatically have to be templated and therefore couldn't be virtual.

nilsdeppe replied:
I checked with @kidder and he likes the idea. @wthrowe is having an `is_equal` function like that described above okay with you?

isaaclegred replied:
To make sure, will this also entail adding in the `NewEosBase` class (or whatever we want to call it)?

wthrowe replied:
Seems fine.  I've seen something like this used before.  I think the only subtlety is that either all the classes that implement `is_equal` have to be final or the method should be private and the base class should have
```c++
friend bool operator==(const Base& a, const Base& b) {
  return a.is_equal(b) and b.is_equal(a);
}
```

isaaclegred replied:
So will the idea to be something like 
```c++
namespace EquationsOfState {
class Base : public PUP::able{
  ...
  virtual is_equal(const Base & b) const = 0;
  friend bool operator==(const Base& a, const Base& b) {
    return a.is_equal(b) and b.is_equal(a);
  }
}
template <bool IsRelativistic>
class EquationOfState<IsRelativistic, 1> : Base {
  public:
  ...
```

nilsdeppe replied:
Sorry, I think just `a.is_equal(b)` should be enough. The checks should be symmetric :)

wthrowe replied:
`is_equal` should be symmetric, but if `a` and `b` have different dynamic types the two calls may be to different functions.  If the types are unrelated, this is still not a problem because the `dynamic_cast` expressions will both fail, but if one is derived from the other (and the more derived type overrides `is_equal`) the less derived `is_equal` can pass while the more derived one fails.  This is admittedly an obscure edge case.  (The more likely edge case is probably a base class with two derived classes, neither of which overrides `is_ready`, but symmetry isn't violated there.)

Perhaps a better solution is `typeid(a) == typeid(b) and a.is_equal(b)`.  That should avoid any subtleties.

nilsdeppe replied:
I like that suggestion :)

nilsdeppe replied:
I think the `TODO` comment can be removed now, right? Or at least once the `operator==` is used in the MTOV class :)

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/Test_TovStar.cpp:
```diff
@@ -39,34 +42,46 @@ void verify_solution(const TovStar& solution, const std::array<double, 3>& x) {
+                      100.0, 2.0),
+                  coord_system});
+    // test_copy_semantics(solution);
+    // TovStar star_copy = solution; // Broken
```
could be restored by adding custom copy semantics that does a `get_clone` I wouldn't worry about it for now, just add a comment that `// Add support for arbitrary EOSes would require custom copy semantics that we haven't needed yet.`

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/Test_TovStar.cpp:
```diff
@@ -21,12 +21,15 @@
 #include "Utilities/Gsl.hpp"
 #include "Utilities/StdArrayHelpers.hpp"
 
+#include "Parallel/Printf.hpp"
```
once done debugging, remove `Printf` includes

nilsdeppe commented on tests/Unit/Evolution/Initialization/Test_SetVariables.cpp:
```diff
@@ -106,8 +107,11 @@ struct SystemAnalyticSolution : public MarkAsAnalyticSolution {
     return vars;
   }
 
+  EquationsOfState::PolytropicFluid<true> equation_of_state_{100.0, 2.0};
```
I'm a bit confused by the two different implementations. The one below returns `equation_of_state_` in the `equation_of_state()` function

isaaclegred replied:
Yeah this didn't make sense...

nilsdeppe commented on src/PointwiseFunctions/Hydro/EquationsOfState/RegisterDerivedWithCharm.cpp:
```diff
@@ -0,0 +1,39 @@
+  using derived_equations_of_state =
+      typename EquationsOfState::detail::DerivedClasses<IsRelativistic,
+                                                        ThermodynamicDim>::type;
+  tmpl::for_each<derived_equations_of_state>(RegisterClassesFunctor{});
```
could also do `Register...` -> 
```cpp
[](auto type_v) { 
    using type = tmpl::type_from<decltype(type_v)>;
    Parallel::register_classes_with_charm<type>();
}
```
Less code and is usually what we do :)

wthrowe replied:
Or even `Parallel::register_classes_with_charm(derived_equations_of_state{});`

isaaclegred replied:
Ah this is nice!

nilsdeppe reviewed (changes requested):
I like the direction this is going. Some suggestions that I think once implemented would basically put this at the place where we can merge it. I'll probably have a few more comments after a second detailed review, but this is already great :D

nilsdeppe reviewed (commented):
I'm happy with the fixes so far. Please go ahead and squash those. I think the copy semantics for TOV can be restored now that you can clone the EOS, right?

isaaclegred commented:
Hi @nilsdeppe I think I've fixed everything from the previous comments, I'm still a bit worried about the Baumgarte-Shapiro example, as it seems like the solution implies there is a place with dlog h / dr > 0 which I think violates the assumptions of the Lindblom TOV solver.  I'm also happy to take a stab at  adding the comparison operators if this is the direction we want to move in!

nilsdeppe commented:
Okay, I'm visiting Fullerton today, but maybe we should meet next week to go through the PR and figure out next steps? I'm back in Pasadena so we can either meet in-person or on a call. Since we probably need to look at code a call might actually be easier, but I'm fine either way. I'll message you on Slack to set up a time :)

isaaclegred commented:
Hey @nilsdeppe, I pushed the changes we talked about today so you can take a look at them (also @wthrowe  if you'd like to see where this is heading).  I think we probably want a test for each EoS that checks 3 things equality of EoSs of the same type with the same parameters, non-equality of EoSs with the same types but different parameters, and non-equality of EoSs with different types.   Let me know what you think (I just realized I forgot to include the `type_id` 
"symmetry checking" in this commit, i'll include it in a fixup)

nilsdeppe commented:
That sounds exactly right to me! The fixup looks good. Could you also add `operator!=`. These are normally just implemented as `not(lhs==rhs)`, so they just farm out to the `operator==` definition so there's only one thing to maintain :)

isaaclegred commented:
@nilsdeppe This should be ready for review again whenever you get a chance!

isaaclegred commented:
@nilsdeppe @wthrowe, feel free to take a look at this when you get a chance!

nilsdeppe commented on src/PointwiseFunctions/Hydro/Tags.hpp:
```diff
@@ -211,4 +213,24 @@ struct MassFlux : db::SimpleTag {
+  static constexpr Options::String help = {"The equation of state to use"};
+};
+
+// TODO : make the equation of state base class generic
```
I think our CI flags `TODO` comments. I'd say you can just remove this chunk of code for now and implement it in a followup PR. No need to hold up the great work here for another feature :D

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/TovStar.cpp:
```diff
@@ -678,9 +706,8 @@ PUP::able::PUP_ID TovStar::my_PUP_ID = 0;
 bool operator==(const TovStar& lhs, const TovStar& rhs) {
   // there is no comparison operator for the EoS, but should be okay as
   // the `polytropic_exponent`s and `polytropic_constant`s are compared
+  // TODO : need an == for the EoS!
```
We have this now, right?

nilsdeppe replied:
Ping on deleting this comment :)

nilsdeppe reviewed (commented):
LGTM! I'm happy for you to squash once @wthrowe has taken a look. Thanks for the ping and sorry I didn't get to this last week! Some minor comments for getting rid of `TODO` comments that CI flags. Don't worry about the initializer list thing... That's a compiler bug. I think clang wants `std::array<int, 2> {{a, b}}` but gcc is okay with `std::array<int, 2> {a, b}` and both I think are okay with `std::array{a, b}`. But `std::vector<int> {a, b}` always only has single brackets. I forget what we had concluded is "technically correct" but in any case, clang being inconsistent with GCC on `std::array`s is annoying and confusing :(

isaaclegred commented:
The  `MagnetizedTovStar::operator==` calls the underlying `TovStar::operator==`, so I think all the both the == TODO's can be removed with the introduction of the EoS ==.  I also added some tests of the CubicSpline copy constructors I added (and added an == for cubic spline to make the tests easier).

wthrowe commented:
I haven't done a full review of this PR yet, so go ahead and squash and I'll look over the whole thing.

I'm going to wait on doing a review until the tests are passing unless you want advice on whatever is causing that.  Feel free to ask in that case.

isaaclegred commented:
> I haven't done a full review of this PR yet, so go ahead and squash and I'll look over the whole thing.
> 
> I'm going to wait on doing a review until the tests are passing unless you want advice on whatever is causing that. Feel free to ask in that case.

My bad @wthrowe, I forgot the EoS was stored as a `unique_ptr` in the tov solutions, I've fixed that up and ran the tests and it looks fine now, but I'm gonna check to make sure the CI agrees.  Once that is done I'll squash and ping you again.

isaaclegred commented:
@wthrowe, the tests are passing now and everything is squashed, feel free to take a look whenever

wthrowe commented on tests/Unit/PointwiseFunctions/Hydro/EquationsOfState/Test_Enthalpy.cpp:
```diff
@@ -43,6 +43,23 @@ void check_exact() {
+      0.0};
+  const auto other_type_eos = PolytropicFluid<true>{100.0, 2.0};
+  CHECK(eos == eos);
+  CHECK(eos != other_eos);
```
Add inequality tests where the low-density part is the same but the high-density part differs, and vice versa.

wthrowe commented on src/PointwiseFunctions/Hydro/Tags.hpp:
```diff
@@ -211,4 +213,17 @@ struct MassFlux : db::SimpleTag {
+namespace OptionTags {
+
+/// The equation of state of the fluid.
+template <size_t ThermoDim = 0, bool IsRelativistic = true>
```
Is 0 a valid thermodynamic dimension?  Is there a reason that these template parameters need to have defaults?

isaaclegred replied:
0 is definitely not a valid thermodynamic dimension, I believe @nilsdeppe mentioned the reason for needing defaults here but I can't remember what it was.  It may have only been needed for the possible extension to a common base class.

nilsdeppe replied:
Oh, that was for when we try to do dimension-agnostic base classes. For this PR I would leave out the defaults :)

wthrowe commented on tests/Unit/Evolution/Initialization/CMakeLists.txt:
```diff
@@ -15,7 +15,7 @@ add_test_library(
   "Evolution/Initialization/"
   "${LIBRARY_SOURCES}"
-  "DataStructures;Domain;Evolution;Utilities"
+  "DataStructures;Domain;Evolution;Utilities;Hydro"
```
Alphabetize.

wthrowe commented on tests/Unit/Helpers/PointwiseFunctions/Hydro/EquationsOfState/TestHelpers.hpp:
```diff
@@ -226,5 +226,13 @@ void check(EosType in_eos, const std::string& python_file_name,
+/// Test that cloning is correct, and that the equality operator is implemented
+/// correctly
+template <class EosType>
+void test_get_clone(std::unique_ptr<EosType> in_eos) {
```
[optional] Take a `const EosType&` (or probably better `const EquationOfState<...>&` with the function templated on the EOS template parameters) as the argument, so callers don't have to give up ownership of a `unique_ptr`.

wthrowe commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_CubicSpline.cpp:
```diff
@@ -86,6 +88,16 @@ void test_cubic_spline(const F& function, const double lower_bound,
+    intrp::CubicSpline interpolant_copy_assigned{{-1.0, 0.0, 1.0, 2.0},
+                                                 {0.0, 0.0, 0.0, 6.0}};
+    interpolant_copy_assigned = interpolant;
+    const auto& interpolant_copy_constructed{interpolant};
```
No `&`.  Currently, this doesn't create a new object.

wthrowe commented on tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/Test_MagnetizedTovStar.cpp:
```diff
@@ -21,6 +21,8 @@
 #include "PointwiseFunctions/InitialDataUtilities/Tags/InitialData.hpp"
 #include "Utilities/GetOutput.hpp"
 
+#include "Parallel/Printf.hpp"
```
Remove.

wthrowe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/Test_TovStar.cpp:
```diff
@@ -39,31 +40,45 @@ void verify_solution(const TovStar& solution, const std::array<double, 3>& x) {
 }
 
 void test_tov_star(const TovCoordinates coord_system) {
+  Parallel::printf("Entered the function which tests");
```
Probably remove, if not add necessary #include.  Same with the other calls to `Parallel::printf`.

wthrowe commented on src/PointwiseFunctions/Hydro/Tags.hpp:
```diff
@@ -211,4 +213,17 @@ struct MassFlux : db::SimpleTag {
+template <size_t ThermoDim = 0, bool IsRelativistic = true>
+struct EquationOfState {
+  using type = std::unique_ptr<
+      EquationsOfState::EquationOfState<ThermoDim, IsRelativistic>>;
```
Am I looking at the wrong class, because it looks like the template parameters are in the wrong order?
```c++
template <bool IsRelativistic, size_t ThermodynamicDim>
class EquationOfState;
```

isaaclegred replied:
No, I think this is a bug which hasn't caused problems because so far this code has only been run with dim= 1

wthrowe commented:
Looks good.  Squash.

isaaclegred commented:
@nilsdeppe do you have any additional comments?

nilsdeppe reviewed (approved):
No, LGTM! @wthrowe would you mind looking at this again? It's been squashed :)
